### ISSUE TITLE
(no-value-for-parameter) on dataclass field attribute C0114

### ISSUE THREAD

**Issue by tweigel-dev** [PyCQA/pylint#5225]
### Bug description

hay,
at working with dataclasses and pylint I noticed that pylint outputs a missleading error for generated attributes of dataclass.
``` python
from dataclasses import dataclass, field


@dataclass
class Person:
    """
    Class implementation to forecast on PRB
    """
    # pylint: disable=missing-function-docstring, too-many-instance-attributes, function-redefined,
    name: str
    last_char: int = field(init=True)

    @property
    def last_char(self):
        return self.name[-1]

    @last_char.setter
    def last_char(self, _):
        pass  # Do nothing, this is a read-only attribute

print(Person("Manfred"))
```

I got *E1120: No value for argument 'last_char' in constructor call (no-value-for-parameter)*

### Command used

```shell
pylint ./bug-dataclasses.py
```

### Pylint output

```shell
E1120: No value for argument 'last_char' in constructor call (no-value-for-parameter)
```

### Expected behavior

No lint error

### Pylint version

```shell
pylint 2.11.1
astroid 2.8.4
Python 3.9.5 (default, May 19 2021, 11:32:47) 
[GCC 9.3.0]
```

**Comment by pgoy-dod** [comment]
Just adding my voice to this ticket to note that this also occurs in subclasses of dataclasses with default values. For example:

```python
from dataclasses import dataclass


@dataclass
class Parent:
    name: str
    num_children: int = 1


@dataclass
class Grandparent(Parent):
    num_grandchildren: int = 1

 # complains with the "no-value-for-parameter" error for num_children, num_grandchildren
grandpa = Grandparent(name="Grampy") 
```

**Comment by plcarman** [comment]
Hitting this as well. The workaround I have for now is to set disable/enable blocks around the problematic code. 

```py
    #pylint: disable=no-value-for-parameter
    grandpa = Grandparent(name="Grampy") 
    #pylint: enable=no-value-for-parameter

```

**Comment by Pierre-Sassoulas** [comment]
Another example from #7059 mentionned by @suryaavala:

```python
# pylint: disable=missing-module-docstring, missing-class-docstring
from dataclasses import dataclass, field


@dataclass
class Mamal:
    name: str
    age: int
    mamal_subgroup: str


@dataclass
class Person(Mamal):
    name: str
    age: int
    mamal_subgroup: str = field(default="primates", init=False)


Person("John", 20)
```

**Comment by DanielNoord** [comment]
Related to https://github.com/PyCQA/pylint/issues/7291
